ILL SANS v2

gvardany opened the pull request:
**Description of work.**
This is the refactoring of ILL SANS suite resulting in 1-2 orders of magnitude of speed-up for monochromatic/kinetic.
The principal change is the concatenation of multiple runs/frames along the x-axis into a single workspace (per detector distance).
Introduces a v2 of `SANSILLReduction`, which supersedes and deprecates the v1.
Introduces a new algorithm `SANSILLMultiProcess`, which deprecates `SANSILLAutoProcess`.
The 2 are tightly coupled, so they had to be written together.
The clean-up of the loader, however, will be done in a separate PR.

**For reviewer**
The diff is rather large, as many child algorithms needed minor interventions.
However, most of the new files are just test data.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**
1. Read the referred issue description, up to the section Workflow included.
2. Build and read the documentation of the 2 new algorithms.
3. Run and time their system tests `systemtest -R SANSILLReduction2` and `systemtest -R ILLSANSMultiProcess`
4. Play with the 2 tests in the  `ILLSANSMultiProcess` system test, while tweaking the parameters.
5. Code review.

<!-- Instructions for testing. -->

Fixes #31943. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

gvardany committed:
mask nans and infs instead of just replacing with 0

gvardany committed:
remove irrelevant beam center offset in default q-binning calculation

Commit:
fix doc test output

Commit:
remove trailing (), check for n_wedges

Commit:
Merge branch '31055_stitch' into sans_midq_qrange

Commit:
use the new stitch in SANSILLAuto

Commit:
Merge branch '31055_stitch' into sans_midq_qrange

Commit:
Merge branch '30343_sansill_directbeam_resolution' into sans_midq_qrange

Commit:
remove blank doc test

gvardany committed:
Merge branch 'master' into sans_midq_qrange

gvardany committed:
enable 4th column by default in ascii exports

gvardany committed:
made outputbinning per sample

Commit:
proper validation for OutputBinning with : as delimiter

Commit:
do not complain on dx if it is requested, but doesn't exist

Commit:
refactored outputting the wedges

Commit:
uncomment the sensitivity outputs

Commit:
refactor outputting panels

Commit:
also set shape_table to empty for not doing wedges for panels

Commit:
check if ws exist before bookkeeping, happens if there is empty token

Commit:
swap det 1 and 2 for D22B starting from cyc 212

Commit:
Merge branch 'master' into sans_midq_qrange

Commit:
chopper att2 for D22 and smarter q interval

Commit:
better error when requesting wrong q range

Commit:
convert att_value to float

Commit:
remove extra flux rescaling

gvardany committed:
support kinetic mode for d11b in the loader

Commit:
implemented kinetic reduction returning a workspace group

Commit:
implemented kinetic (mono) reduction

Commit:
guard for workspace group in needs_processing

Commit:
handle absent or present 2nd monitor, store durations as the 2nd mon

Commit:
round wavelength for kinetic, require alphanum for output

Commit:
set the 2nd monitor (a.k.a durations) errors to 0

Commit:
Merge branch 'master' into sans_midq_qrange

gvardany committed:
Merge branch '31055_stitch' into sans_midq_qrange

gvardany committed:
added a unit test for kinetic and D33 VTOF

gvardany committed:
fix the regression for d16 omega scan, add unit test

gvardany committed:
updated sys test reference for SANSILLReduction

gvardany committed:
update system test reference for autoprocess, chunk 1

gvardany committed:
fix the test name

gvardany committed:
update system test references for autoprocess (chunk 2)

gvardany committed:
fix the validation ws name in iphiq

gvardany committed:
Merge branch 'sans_midq_qrange' into ill_sans_v2

gvardany committed:
PyInit of SANSILLMultiProcess

gvardany committed:
adjust the test of SaveAscii2 when there is no dx

gvardany committed:
adjust the doc test

gvardany committed:
Merge branch 'sans_midq_qrange' into ill_sans_v2

gvardany committed:
input validations

gvardany committed:
Merge branch '31055_stitch' into sans_midq_qrange

gvardany committed:
Merge branch 'sans_midq_qrange' into ill_sans_v2

gvardany committed:
the skeleton of the pyexec

gvardany committed:
inception of v2 of SANSILLReduction and SANSILLCommon

gvardany committed:
fixed flake8 minor issues

gvardany committed:
normalisation and dead time corr

gvardany committed:
split out normalise to time

gvardany committed:
the skeleton of the reductions

gvardany committed:
apply empty beam, transmission and container

gvardany committed:
apply container, water, solvent

gvardany committed:
implemented process beam

gvardany committed:
calculate transmission

gvardany committed:
delete the solid angle workspace

gvardany committed:
a note about loops in the comments

gvardany committed:
remove wavelength validator from parallax correction

gvardany committed:
deprecate the SANSILLReduction-v1 and SANSILLAutoProcess

gvardany committed:
apply dead time corr also for kinetic

gvardany committed:
remove wavelength requirement in calculate flux, rebin only if TOF

gvardany committed:
added linearize axis to conjoin

gvardany committed:
ammend calculate dynamic range to take wavelength from sample logs

gvardany committed:
remove the validator from apply transmission correction

gvardany committed:
Merge branch '31055_stitch' into sans_midq_qrange

gvardany committed:
Merge branch 'sans_midq_qrange' into ill_sans_v2

MathieuTillet committed:
Discrete move for integration bar

gvardany committed:
Merge branch '31055_stitch' into sans_midq_qrange

gvardany committed:
Merge branch '31055_stitch' into sans_midq_qrange

gvardany committed:
Merge branch 'sans_midq_qrange' into ill_sans_v2

gvardany committed:
fixed d11 multi transmission ref following a bug fix in stitch

gvardany committed:
fixed D16 gamma scan ref after a fix in stitch

gvardany committed:
updated reference of d11 abs scale test

gvardany committed:
fixed the inputs in drill process test to match the other sys test

gvardany committed:
scale by the flux factor before comparison, due to commit f25a6cd5

gvardany committed:
Merge branch '31055_stitch' into sans_midq_qrange

gvardany committed:
Merge branch 'sans_midq_qrange' into ill_sans_v2

MathieuTillet committed:
Fix integration range last bin bug

When integrating data using the instrument viewer, the last bin from the
data was systematically excluded - because accumulate does not use the
data from the provided upper bound

MathieuTillet committed:
Control the integration slider with spin boxes

When the data is disrete, the slider is controlled by spin boxes
defining min and max values.

MathieuTillet committed:
Change discrete criterion

Choose discrete mode in the integration slider of the instrument viewer
if all the bins are common and integers

gvardany committed:
Merge branch '31055_stitch' into sans_midq_qrange

gvardany committed:
doxygen, removed unneeded include statements

gvardany committed:
added a test for D22B cycle 212 onwards

gvardany committed:
added release notes for extended algorithms

gvardany committed:
Merge branch 'master' into sans_midq_qrange

gvardany committed:
Merge branch 'sans_midq_qrange' into ill_sans_v2

gvardany committed:
cherry-picked the angle offsets for parallax correction

gvardany committed:
Merge branch 'master' into sans_midq_qrange

gvardany committed:
fixed the doxygen param issues

gvardany committed:
Merge branch 'sans_midq_qrange' into ill_sans_v2

gvardany committed:
incorporated concatenation to LoadAndMerge

gvardany committed:
assume acq time is in 2nd monitor also for standard mono

gvardany committed:
correct linearize axis in conjoin and inject blanks in reduction

gvardany committed:
load first to a hidden workspace

gvardany committed:
broadcast transmissions and thicknesses

gvardany committed:
add aux flux output ws, delete tmp ws in calculate flux

gvardany committed:
do not rebin in calculate efficiency, apply sa to all above container

gvardany committed:
fix a typo in doc string

gvardany committed:
Merge branch 'master' into ill_sans_v2

gvardany committed:
add missing apply_flux to reduction

gvardany committed:
fix deletion of loaded workspaces after merging

gvardany committed:
fix flux ws check in validateInputs

gvardany committed:
first system test for reduction v2 for D11

gvardany committed:
added a system test for D22

gvardany committed:
added a system test for D22 multiple water for sensitivity case

gvardany committed:
update the parallax formula from CD giving smoother results

gvardany committed:
Merge branch 'master' into ill_sans_v2

gvardany committed:
simplify if/else in calculate efficiency

gvardany committed:
update the parallax formula (from CD) for D11B

gvardany committed:
check multiple mono from D11B, including with holes

gvardany committed:
broadcast only if needed, fixed a typo in comments

gvardany committed:
introduce temp preprocess, handle blanks better

gvardany committed:
fix putting acq time for revent

gvardany committed:
polish the d16 gamma scan step case

gvardany committed:
fixed a typo in the logic expression

gvardany committed:
remove wavelength unit validator from integration

gvardany committed:
improve docstrings in integration algorithm

gvardany committed:
fixes to get d33 ltof work again

gvardany committed:
move empty token to commons, give an example in comment

gvardany committed:
rearranged the functions into logical groups

gvardany committed:
introduced progress reporting

gvardany committed:
implemented transmission repetition logic

gvardany committed:
add dummy time log for kinetic to avoid error in mergeruns

gvardany committed:
move a method to the helper block

gvardany committed:
apply beam center corr before flux and transmission calc

gvardany committed:
fix apply beam center when doing transmission

gvardany committed:
correctly enable/disable the beam and tr beam radii

gvardany committed:
improve docstrings of some methods

gvardany committed:
cleanup unneeded workspace outputs

gvardany committed:
implemented process transmissions

gvardany committed:
implement process sample

gvardany committed:
produce and apply the sensitivity map

gvardany committed:
encorporated the integration to multi process

gvardany committed:
enabled wedges and panels

gvardany committed:
group methods by family

gvardany committed:
remove sample and transm lists from the class variables

gvardany committed:
implemented combined sensitivity generation

gvardany committed:
incorporated the new stitch as the final step

gvardany committed:
implemented stitching of the wedges

gvardany committed:
convert iqs, iqw, iqp and stitched outputs back to distribution

gvardany committed:
move the need* to commons

gvardany committed:
rename iq, iqw, iqp and realspace

gvardany committed:
streamlined rename logic

gvardany committed:
set the accumulable acq time in duration log only

gvardany committed:
temporarily disable the WIP tests

gvardany committed:
Merge branch 'master' into ill_sans_v2

gvardany committed:
exclude the 3rd test temporarily too

gvardany committed:
add doxygen comment for new argument in parallax correction

gvardany committed:
package also sample transmissions

gvardany committed:
mask non-finite before integration

gvardany committed:
implemented sample thickness: [user, nexus]

gvardany committed:
included thickness validation in validateInputs

gvardany committed:
implemented setting of sample name axis for iq, iqw, iqp and stitched

gvardany committed:
split generation and replacement of the axis, also cached it

gvardany committed:
implemented progress reporting

gvardany committed:
some fixes from full a complete reduction tests

gvardany committed:
Merge branch '32032_connection_control_workspacecalculator' into ill_sans_v2

gvardany committed:
Merge branch 'master' into ill_sans_v2

gvardany committed:
fix default values for beam radius and tr beam radius

gvardany committed:
fixed the remove_repeated function to ensure stable sort

gvardany committed:
Merge branch '32032_connection_control_workspacecalculator' into ill_sans_v2

gvardany committed:
set resolution as dx to all the spectra

gvardany committed:
added more docstrings

gvardany committed:
use version 1 in old tests, use duration vs timer

gvardany committed:
use v1 in auto-process algorithm

gvardany committed:
version lowercase in copyProperties

gvardany committed:
added unit tests for the new multiprocess algorithm

gvardany committed:
fix the integration unit test

gvardany committed:
use v1 in parameter scan and its test

gvardany committed:
fix sample logs model test

gvardany committed:
use v1 also in docs tests

gvardany committed:
remove repeate keyword argument in test

MathieuTillet committed:
Refactor and simplify logic

gvardany committed:
implemented loading of kinetic for D22B

MathieuTillet committed:
Integrate on index on discrete data X bins

Rather than integrating on the X bins values, if the axis is discrete,
use the indices on the slider. Solve case where the axis has discrete
bins with spacing greater than 1.

gvardany committed:
fixed the check for wavelength log

MathieuTillet committed:
Fix occasionnal segfault due to workspace deletion

The workspaces being unlocked before the the finished notification is
sent, they could have been deleted too early, leading to a segfault as
poco notification is about a workspace that does not exist anymore.

MathieuTillet committed:
Add release notes

MathieuTillet committed:
Fix instrument view test

Since the values in the test are integers, the workspace is considered
to have a discrete bin axis and thus the integration range is not the
bin value but its index.

gvardany committed:
set strong focus on the slider to enable key press on macOS

gvardany committed:
change absolute scale tests to use v1 of the reduction

gvardany committed:
Merge branch 'master' into ill_sans_v2

MathieuTillet committed:
Start discrete integration in first bin

gvardany committed:
support Revent mode as input, change mon normalisation to ws index

gvardany committed:
[TMP] place D22B detector in the algorithm

gvardany committed:
Merge branch '31943_discrete_integration_slider' into ill_sans_v2

gvardany committed:
Merged branch 31747_iview_event_rebin

gvardany committed:
updated systest reference due to the parallax correction

gvardany committed:
correct setting distribution flag before and after stitch in v1

gvardany committed:
fix list argument

gvardany committed:
Merge branch 'master' into ill_sans_v2

gvardany committed:
Merge branch '31943_discrete_integration_slider' into ill_sans_v2

gvardany committed:
detector placement for D11B events

gvardany committed:
remove deprecation warnings until drill v2 is available

gvardany committed:
Merge branch 'master' into ill_sans_v2

gvardany committed:
remove remnant of merge conflict resolution

gvardany committed:
copy wavelength from v1 of the algorithm in auto process

gvardany committed:
use stl algorithm instead of raw loop

gvardany committed:
guard for a type in wavelength res for D22

MathieuTillet committed:
Propagate wavelength to all loaded files

Make it so the wavelength parameter is given to all files loaded by
sansillautoprocess, not only the sample

MathieuTillet committed:
resolved merge conflict with between wavelength and version

gvardany committed:
Merge remote-tracking branch 'mantill/d16_wavelength_issue' into ill_sans_v2

gvardany committed:
enable basic systemtests for the reduction v2

gvardany committed:
change teardown to cleanup in sys tests

gvardany committed:
discard remnant for merge conflict resolution in xintcontrol

gvardany committed:
apply parallax also for old D22, with new formula

gvardany committed:
updated sys test references for old D22

gvardany committed:
update sys test ref for old algorithm too

gvardany committed:
added unit tests for the reduction2 for monochromatic case

gvardany committed:
added a system test for D11B mono, 3 distances

gvardany committed:
added system test for D22B reduction in absolute scale

gvardany committed:
fix cppcheck issues

gvardany committed:
added system test for reduction 2 for D33 mono

gvardany committed:
added system test for multiprocess for monochromatic

gvardany committed:
Merge branch 'main' into ill_sans_v2

gvardany committed:
remove obsolete cppcheck suppressions

gvardany committed:
remove return to unused variable

gvardany committed:
fixed name axis for tr workspace, added new systest for kine

gvardany committed:
added test data for the kinetic sys test

gvardany committed:
added documentation for reduction v2

gvardany committed:
added documentation page for multi process

gvardany committed:
Merge branch 'main' into ill_sans_v2

gvardany committed:
added sys test in multiprocess for D33 mono, improved algo comments

gvardany committed:
remove D16 from the comment on supported instruments

martyngigg commented:
@gvardany This has developed some conflicts. Are you still aiming this 6.3?

gvardany commented:
Well, in an absence of a reviewer, I guess not any more.

gvardany committed:
added system test in reduction v2 for D33 VTOF

gvardany committed:
Merge branch 'main' into ill_sans_v2

gvardany committed:
moved the release note to 6.4

gvardany committed:
added a system test for reduction v2 for D33 LTOF

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLReduction2.py:
```diff
@@ -0,0 +1,952 @@
+                                                     optional=PropertyMode.Optional),
+                             doc='The name of the input empty beam flux workspace.')
+
+        self.setPropertySettings('FluxWorkspace', water_solvent_sample)
```
In the UI, I see that FluxWorkspace is disabled when ProcessAs=Transmission. Should this be enabled? I was keying in a few of of the calls to SANSILLReduction that were in one of the system tests when I noticed this.

DannyHindson commented on docs/source/algorithms/SANSILLMultiProcess-v1.rst:
```diff
@@ -0,0 +1,44 @@
+Optionally, it can also perform anisotropic integration with azimuthal wedges or hand-drawn sectors in the instrument viewer.
+If requested, it can also produce separate :math:`I(Q)` curves per detector panel.
+It supports standard monochromatic, kinetic and rebinned event modes.
+Makes use of :ref:`SANSILLRedcution <algm-SANSILLReduction-v2>` and :ref:`SANSILLIntegration <algm-SANSILLIntegration>`.
```
There's a typo in the link SANSILLRedcution

DannyHindson reviewed (changes requested):
I'll continue on Monday but I've had an initial look around. Not found v much so far. I've built the docs and had a look at the workflows - they all look fine apart from 1 minor typo.

I think the changes to the various child algorithms probably need a release note in Framework.rst in addition to the sans.rst release note you've already done:
- CalculateDynamicRange (the doc page for this algorithm also says it requires an input workspace with x unit=wavelength)
- ConjoinXRuns
- ParallaxCorrections
- ApplyTransmission
- LoadAndMerge

I ran the system tests locally and timed them as requested in the testing instructions. They were both ~2 mins. What was the aim of this - just to check they're not too long or should it be possible to do a before\after comparison and get some idea of the performance speed up?

There's one enable rule on the algorithm properties needs adjusting as well I think (FluxWorkspace)

gvardany committed:
fix the criteria for flux workspace, add release notes for child algos

gvardany commented:
I've addressed your comments so far, good catches!
For the running time of system tests, the point was just to confirm they are in the minute range.
Benchmarking has been done on a complete experiment dataset.

gvardany committed:
fallback to zero as wavelength value for TOF

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLMultiProcess.py:
```diff
@@ -0,0 +1,990 @@
+            flat_field_ws = self.load_flat_field(d)
+            solvent_ws = self.load_solvent(d)
+            sens_ws = self.load_sensitivity(d)
+            sample_tr_ws = transmissions[self.tr_index(d)]['SampleTransmission'] if transmissions else ''
```
If I run the algorithm with SampleRunsD1 populated and SampleTrRunsW2 populated (but SampleTrRunsW1 empty) I get an error from this line because transmissions =[{}] and this evaluates to true because it's a non-empty list

gvardany replied:
Transmission runs as well have to be populated from W1 onwards. If there is only one wavelength, they have to be filled at W1, I'll add a validation for this, too.

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLMultiProcess.py:
```diff
@@ -0,0 +1,990 @@
+
+        #=============================INPUT FILES/WORKSPACES==============================#
+
+        self.declareProperty(name='SensitivityMap', defaultValue='',
```
If I enter a nonsense string for any of these "input files\workspaces" properties in the UI and run the algorithm, the dialog closes but an error is reported from the LoadNexusProcessed algorithm. Could these be validated in the dialog?

gvardany replied:
Ideally there should be File/Workspace property to handle all of that. But in the absence thereof, maybe I could use the `FileFinder` to validate that.

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLReduction2.py:
```diff
@@ -0,0 +1,952 @@
+            Scale(InputWorkspace=ws, Factor=1/thicknesses[0], OutputWorkspace=ws)
+        else:
+            if thicknesses[0] < 0:
+                thicknesses = mtd[ws].getRun()['SampleThickness'].value.split(',')
```
I think the code needs to check that there is an entry called 'SampleThickness' in the sample logs. If I specify SampleThicknessFrom Nexus or enter a negative sample thickness in the UI I can make this line generate an error

gvardany replied:
Added a check for this.

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLMultiProcess.py:
```diff
@@ -0,0 +1,990 @@
+        #===================================I(Q) OPTIONS==================================#
+
+        self.declareProperty(name='OutputBinning', defaultValue='',
+                             doc='Output binning for each distance( : separated list of binning params).')
```
If OutputType=None do any of these I(Q) options apply? If no then could they be disabled?

gvardany replied:
That flag was a leftover, I removed that.

DannyHindson reviewed (changes requested):
I've been through it all now. I've added a few minor things - mostly missing validations when running the new workflow algorithms from the UI with funny inputs.
The overall change and new approach of concatenation of multiple runs/frames along the x-axis looks fine to me

gvardany commented on Framework/PythonInterface/plugins/algorithms/CalculateFlux.py:
```diff
@@ -41,13 +37,16 @@ def PyInit(self):
-        width = (max_wavelength-min_wavelength) / blocksize
-        params = [min_wavelength, width, max_wavelength]
-        rebinned = Rebin(InputWorkspace=input_ws, StoreInADS=False, Params=params)
+        if input_ws.getAxis(0).getUnit().unitID() == 'Wavelength' and not input_ws.isCommonBins():
```
Regarding the 2, if the unit is dimensionless, it should never be ragged, as this doesn't make sense.

DannyHindson replied:
Re 2), the scenario I was thinking of was if someone ran this algorithm on a ragged workspace with units of "TOF" for example. Before these changes the workspace would have been rebinned and the call to GroupDetectors would have succeeded (even though the variable names refer to wavelength the Rebin would still work) . After these changes the Rebin is skipped and GroupDetectors would fail I think

So not a problem for your workflow but small chance it might affect someone else. I admit this is a bit obscure though so perhaps not very likely to happen

gvardany replied:
Ok, I see, I switched the condition to be `!='Empty'` instead of `='Wavelength'`. It's only in case of empty units that it must not be ragged.

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLAutoProcess.py:
```diff
@@ -473,6 +473,7 @@ def PyExec(self):
 
         GroupWorkspaces(InputWorkspaces=outputSamples,
                         OutputWorkspace=self.output)
+        self.set_distribution(outputSamples)
```
Yes, minor fixes were needed.

gvardany committed:
more tests and comments for conjoin and loadandmerge

gvardany committed:
catch validation err in validateInputs, guard for empty sample runs

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLMultiProcess.py:
```diff
@@ -0,0 +1,990 @@
+    def _set_n_samples(self):
+        '''Sets the number of samples based on the inputs of the sample runs in the first non-empty distance'''
+        for d in range(self.rank):
+            if self.getPropertyValue(f'SampleRunsD{d+1}'):
```
Indeed, here the loop was supposed to run over the maximum `N_DISTANCES`.  
Still, the sample runs must be filled from 1 upwards, with no gaps. 
I'll add validation for this.
The feature to specify gaps is the placeholder run number `0`, but one can't leave the input completely empty.

gvardany committed:
change the condition on unit to be !=empty

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLCommon.py:
```diff
@@ -0,0 +1,260 @@
+        return run['Detector 1.det1_calc'].value
+    if instrument == 'D33':
+        return run['Detector.det2_calc'].value
+
```
Agreed, it's better to raise the error early.

gvardany committed:
raise error if main detector distance is not found

gvardany committed:
Merge branch 'main' into ill_sans_v2

gvardany committed:
remove output type property

gvardany committed:
fix sample thickness log name, check if it exists

gvardany commented:
Thanks for the review, I think I addressed all of the comments.

DannyHindson commented:
There's an error now when I open the UI for the SANSILLMultiProcess algorithm. Looks like the same problem is causing the doc build to fail on a couple of the Jenkins jobs.
I think it's caused by the fact you've removed the OutputType property. This is still referred to in the "EnabledWhen" rule for the DefaultQBinning property which is copied from the SANSILLIntegration algorithm:
```
        self.declareProperty(name='DefaultQBinning', defaultValue='PixelSizeBased',
                             validator=StringListValidator(['PixelSizeBased', 'ResolutionBased']),
                             doc='Choose how to calculate the default Q binning.')
        self.setPropertySettings('DefaultQBinning', EnabledWhenProperty(output_iq, output_iphiq, LogicOperator.Or))
```

gvardany committed:
add output type back, with only one option - I(Q)

gvardany commented:
> There's an error now when I open the UI for the SANSILLMultiProcess algorithm. Looks like the same problem is causing the doc build to fail on a couple of the Jenkins jobs. I think it's caused by the fact you've removed the OutputType property. This is still referred to in the "EnabledWhen" rule for the DefaultQBinning property which is copied from the SANSILLIntegration algorithm:
> 
> ```
>         output_iq = EnabledWhenProperty('OutputType', PropertyCriterion.IsEqualTo, 'I(Q)')
>         output_iphiq = EnabledWhenProperty('OutputType', PropertyCriterion.IsEqualTo, 'I(Phi,Q)')
>         output_iqxy = EnabledWhenProperty('OutputType', PropertyCriterion.IsEqualTo, 'I(Qx,Qy)')
> 
>         self.declareProperty(name='DefaultQBinning', defaultValue='PixelSizeBased',
>                              validator=StringListValidator(['PixelSizeBased', 'ResolutionBased']),
>                              doc='Choose how to calculate the default Q binning.')
>         self.setPropertySettings('DefaultQBinning', EnabledWhenProperty(output_iq, output_iphiq, LogicOperator.Or))
> ```

Indeed, the copyProperties somehow didn't like property criteria based on another property that wasn't copied, even if its value was equal to its default. Anyway, I had to add `OuputType` back to the master algorithm, just with a single option I(Q).

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLCommon.py:
```diff
@@ -0,0 +1,261 @@
+    if property_value:
+        run_number = get_run_number(property_value)
+        ws_name = run_number + '_' + process_reduction_type
+        if mtd.doesExist(ws_name):
```
No, this is not supposed to work if a critical setting, such as the normalization changes. Supporting it will make things far more complicated. When changing this, one must clear the workspaces and start over. I'll add a note about this in the documentation.

gvardany committed:
improve the doc of multi-process algorithm

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLMultiProcess.py:
```diff
@@ -0,0 +1,1029 @@
+        all_outputs = []
+        for l in range(self.lambda_rank):
+            transmissions = self.process_all_transmissions_at_lambda(l)
+            self.progress.report((l+1)*self.n_samples, 'Calculated transmissions for wavelength index {l+1}')
```
Was missing `f` prefix, good catch.

gvardany committed:
put prefix f for the f-strings in progress reports

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLReduction2.py:
```diff
@@ -0,0 +1,955 @@
+        FindCenterOfMassPosition(InputWorkspace=ws, DirectBeam=True, BeamRadius=radius, Output=centers)
+        beam_x = mtd[centers].cell(0,1)
+        beam_y = mtd[centers].cell(1,1)
+        AddSampleLog(Workspace=ws, LogName='BeamCenterX', LogText=str(beam_x), LogType='Number', LogUnit='meters')
```
I added a validation for this. Most of D33 data have zero monitor counts, so I'll also switch back the default to be the time.

gvardany committed:
switch default normalisation to time, check for zero monitor counts

gvardany committed:
add unit tests for tof and kinetic for reduction v2

gvardany committed:
fixed the broken mon normalisation in test

gvardany commented:
@DannyHindson I guess that's it for the last round of comments.

DannyHindson reviewed (changes requested):
Thanks for the changes. I've had another look around. I found one more area where I wonder if some extra validation would help. I'll probably stop testing obscure\strange combinations of inputs following at this stage because I'm getting diminishing returns and perhaps some of these cases aren't likely to be ever used for real.

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANSILLReduction2.py:
```diff
@@ -0,0 +1,970 @@
+            RenameWorkspace(InputWorkspace=tmp, OutputWorkspace=tmp+'_0')
+            GroupWorkspaces(InputWorkspaces=[tmp+'_0'], OutputWorkspace=tmp)
+        if isinstance(mtd[tmp], WorkspaceGroup):
+            self.setup(mtd[tmp][0])
```
Yes that's fine. There's already a comment in there about not mixing different numbers of frames ("It is understood that in this case the number of frames must match") so maybe just add a comment about mixing modes

gvardany committed:
add a sentence and a comment about input runs being of similar origin

gvardany committed:
added a test to check kinetic calibrants not allowed

gvardany committed:
added tests for checking the order of filled sample and tr runs

gvardany commented:
@DannyHindson I addressed those comments.

DannyHindson reviewed (approved):
All looks good now. Thanks for the changes.

Do you want to keep all of the commits or can this be squashed? Gatekeeper can do this when merging

gvardany commented:
I think it is better to keep as one commits, in the unlikely case if something in the child algorithms needs reverting.
